package.json

```diff
@@ -50,13 +50,15 @@
 		"sprintf-js": "^1.1.1",
 		"typescript": "4.1.3",
 		"wd": "^1.11.1",
-		"yarn": "^1.22.10"
+		"yarn": "^1.22.10",
+		"block-experiments": "Automattic/block-experiments#add-mobile-app-support"
```

Comment:
👋 Enej! I'm generally weary of using dependency refs that are "moving" and can lead to un-reproducible builds when new commits have landed in the referenced branch.

I have the feeling that maybe this is an attempt to bypass another problem though, the one about how to put the proper/desired commit hash in here anyway, in particular how to maintain it as one works on the branch. 🤔

Reply:
I can't speak for Enej's intentions here, but my undestanding of a workflow here is that a package reference could be "floating" on an in-progress PR, but that ultimately, hashes, or better yet, tags would be used for whatever is committed in the dependency, and updated w/ semver accordingly in the related PR. Also, assuming the configuration for npm dependencies can be made to work, I think if this "floating" approach would be accompanied by lock file versioning, which could serve as a signal too, for reviewers to run `npm i` (similar to how we currently run `git submodule update`).

Reply:
Indeed, the process moves to maintaining the lock file instead of the package.json file. I feel that that workflow is more fragile, especially since the lock file can actually get out of hand with all the changes that can happen in it. Having to rely on the lock file becoming essentially a source file is awkward, but I understand it might just come down to personal preference.

Reply:
@hypest Thanks for bringing your concerns up. I also share them. :) 

This PR was definitely an experimental approach. Trying to see what it would take to make things work. So I thought it would be ok to use the branch name for now. 

Like @mkevins suggested I think we would want to manage the code with a tag or a commit hash.

An example of how this is currently done can be found in calypso. 
https://github.com/Automattic/wp-calypso/blob/trunk/apps/editing-toolkit/package.json#L143
In order to update the package. The process seems to be that a new tag gets created on the and then a new commit in the calypso repo. [For example](https://github.com/Automattic/wp-calypso/commit/b3d3062e61115fe35a55f577c8eaa9d86cfaa254#diff-6fe967ae8f9263d474fe549abca821d4e9835e8488601853df9a1a26ace570af)

The current release process of the block-experiments repo already does use tags and has a build process to create the different plugins that eventually end up on WordPress.org. I think it would be valuable to see if we can integrate into this build process to create new versions of the plugin that then gets included in our repo. 

We could even try to use the .org repo directly by using links such as 
https://downloads.wordpress.org/plugin/layout-grid.1.5.zip. 

The first step there would be to add a `package.json` in that zip. I just tried adding 
"layout-grid": "https://downloads.wordpress.org/plugin/layout-grid.1.5.zip" to our `packages.json`
and got the following error. 

```
npm ERR! code ENOPACKAGEJSON
npm ERR! package.json Non-registry package missing package.json: layout-grid@https://downloads.wordpress.org/plugin/layout-grid.1.5.zip.
npm ERR! package.json npm can't find a package.json file in your current directory.
```

However, some of the concerns about properly versioning things are definitely a concern. 
I am not sure how to best address them besides more developer education. Making sure that we are on the same page about how we release software. Such as creating releases and versioning things. 
For example, code published in NPM doesn't necessarily reflect a particular Github tag.

Reply:
> accompanied by lock file versioning, which could serve as a signal too, for reviewers to run `npm i`

👋 @mkevins , just to be sure, do you happen to mean `npm ci` (instead of just `i`) there so to ensure the lock file is enforced?

> Making sure that we are on the same page about how we release software.

Good point. I'd still file my concerns under the "development flow" actually, not the "release flow" per se, though it's true that when dealing with libraries spread out in different repos/infrastructures, there's an element of "releasing" that happens anyway during developing. Generally, I'm leaning more towards having non-standard release flows if that serves the development flows better, since these days development still happens way more oftenly than releasing. When the project matures enough that development has "concluded" then we should lean to making releasing easier instead. Again, this is just my thought process, it doesn't mean we should abandon experimenting like here in this PR 👍.

Reply:
:wave: @hypest 

> just to be sure, do you happen to mean `npm ci` (instead of just `i`) there so to ensure the lock file is enforced

Well, it depends. When "mapping" git submodules and npm dependencies conceptually, I look at the lock file as similar to the git submodule hash, perhaps garnering a bit more semantic information from the dependency graph. For reproducibility (especially CI) `npm ci` is perfect, since it treats the lock file as the source of truth.

`npm i`, on the other hand, allows the versions to "float", which could be handy if used with that intention (or better yet, `npm up <some-package>`), but when used without deliberation can be a source of unwanted "drift". For example, different versions of `npm` may resolve the graph in slightly different ways from a clean state, and so things like `.nvmrc` and `nvm use` become helpful for consistency. In the case where any versions are updated, this would be signalled by diff in the lock file.

Reply:
Oh I see. Yeah, intentionality is key. I'd expect folks to invoke `npm i` exactly when the intention is to allow for upgrades.

I'm trying to think of the PR review flow and who and why would want to upgrade a dep while the PR is in progress. I understand that the author has of course perfect intention to upgrade when needed. The reviewer most probably not, unless they are helping with updating the PR, otherwise I'd assume the reviewer will always go for `npm ci` to pull what the author intents to merge and nothing more.

In that view, the lock file does indeed play the same role as the submodule hash; it will actually be a hash pointer itself too for most of the PR's lifecycle :).

I guess my concern with the lock file is that it tends to pick up changes in other (3rd party) deps, obscuring the intentional "dep bump" change. That is, it can make the reviewing harder. Contrast that to a submodule bump where the change is super isolated and easy to spot.

I wonder if we can work around that "busyness" of the lock file by using https://docs.npmjs.com/cli/v7/commands/npm-shrinkwrap to pinpoint exactly the dependency versions/bumps we want for the particular deps. Like, essentially we want some deps to be closer to the project than the average dep and subdep, and perhaps maintaining a list for that special set of "1st connection" deps in the shrinkwrap file is a better indication of intentionality?

I still kinda wonder what shortcoming of the git submodules we are trying to sidestep but, yeah, what do y'all think about that shrinkwrap idea above?

Reply:
> I'm trying to think of the PR review flow and who and why would want to upgrade a dep while the PR is in progress. I understand that the author has of course perfect intention to upgrade when needed. The reviewer most probably not, unless they are helping with updating the PR, otherwise I'd assume the reviewer will always go for `npm ci` to pull what the author intents to merge and nothing more.

There are a lot of options here. The one most similar to the current submodule hash update would be to use `npm ci`. In that flow, all dependencies are "locked" in place (transitive deps as well as the package under development). In that case, the PR author will need to always update the lock-file in the associated `gutenberg-mobile` PR, just as we do currently with the submodule hash.

Another idea is to allow the commitish to float, and the source of truth be "upstream". In that case, the PR author need not update the `gutenberg-mobile` PR (unless there are source changes there), and instead the experiments branch would be referenced. In that case, to pull the latest, the reviewer would run `npm up block-experiments`. This would also bring in updates to transitive dependencies for that package, if they are specified as floating there, and those changes (if they exist) will be visible in the lock-file diff. This could be desirable, for example, if the direct dependency (in this case `block-experiments`) specified only patch versions to float (i.e. `~` and not `^`), as it would bring bug fixes and sometimes important security updates.

On the other hand, exact versions could be specified in the direct dependency as well. Since this is under our control, this offers a different kind of flexibility, which may serve us well in scaling, since the choice to allow a version to float is very granular. I can imagine a scenario where the direct dependency has a related (transitive) dependency under development, and in that case, it could be helpful to have everything locked _except_ that dependency. The depth of this is entirely configurable on a PR by PR basis, which is an attractive quality.

Regarding `npm-shrinkwrap.json`, that is also a configuration to keep in mind. As I understand it, when we have specified an _exact_ version in our `package.json` file (i.e. a direct dependency), we are still not guaranteed a deterministic outcome via `npm i`, since the transitive dependencies that package brings (hoisted or not) can be specified as floating in the `package.json` of the direct dependency. While, in this case, this is entirely in our control, shrinkwrap offers locking of the transitive dependencies, even without locking them with exact versions in the `package.json` of the direct dependency. I see this as more of a feature for library development though, to ensure a library version is deterministic, which is crucial for release versions.

> I still kinda wonder what shortcoming of the git submodules we are trying to sidestep

I don't think we're trying to sidestep any shortcomings with git submodules; that approach has been working well. I think the idea here is to explore another avenue for dependency management, which offers some interesting options / flexibility which may also serve us well as we scale.